shark_turbine/kernel/wave/codegen.py

```diff
@@ -239,6 +240,68 @@ def _compute_offset(indices: list[int], strides: list[int]) -> int:
     return int(sum(i * s for i, s in zip(indices, strides)))
 
 
+def _get_symbolc_shape(node: fx.Node) -> tuple[IndexExpr]:
+    return get_custom(node).type.symbolic_shape
+
+
+def _construct_gather_scatter_indices(
+    emitter: WaveEmitter,
+    symbolc_shape: tuple[IndexExpr],
+    index: tuple[IndexExpr],
+    mapping: IndexMapping,
+    elements_per_thread: int,
+    is_read: bool,
+) -> tuple[OpResult, OpResult, OpResult]:
+    # Apply symbolc_shape order to indices, e.g. if original mapping is
+    # {M: iter(0), N: iter(1)} and symbolc_shape is (N, M), result will
+    # be (iter(1), iter(0))
+    if is_read:
+        assert (
+            mapping.is_output_identity()
+        ), "non-identity output mapping is not supported yet"
+        index_mapping = mapping.map_input_indices(symbolc_shape)
+    else:
+        assert (
+            mapping.is_input_identity()
+        ), "non-identity input mapping is not supported yet"
+        index_mapping = mapping.map_output_indices(symbolc_shape)
+
+    iters = mapping.iters
+
+    # As we only support identity input/output mapping for now, we can directly
+    # substitute iterators with corresponding expanded index.
+    subs = [(sym, expr.start) for sym, expr in zip(iters.keys(), index.values())]
+
+    # Contruct input/output index, substituting iterators in input mapping with
+    # expanded index.
+    result_index = {key: m.subs(subs) for key, m in zip(symbolc_shape, index_mapping)}
+
+    strides = strides_from_symbolic_shape(IndexingContext.current(), symbolc_shape)
+    offsets = []
+    subs = [(sym, 0) for sym in iters.keys()]
+    for i in range(elements_per_thread):
+        # Update most-minor dim, i.e. in case of identity mapping it will
+        # be quivalent to just vector.load
+        subs[-1] = (subs[-1][0], i)
```

Comment:
Shouldnt the choice of which $index{..} to be updated be determined based on the mapping? Seems like this is hardcoded to the innermost dim.

Reply:
Actually, No. `index_mapping` on next `[int(i.subs(subs)) for i in index_mapping]` line is already updated according to mapping. It will be more complicated if we have both input and out non-identity mapping, bit I don't support this case yet.

Reply:
Sounds good, thanks!